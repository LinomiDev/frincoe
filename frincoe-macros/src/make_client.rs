use proc_macro2::TokenStream;
use quote::quote;
use syn::parse::{Parse, ParseStream};
use syn::{Ident, LitStr, Path, Token, TraitItem, Type};

use crate::helpers::read_trait;

struct ClientArgs {
    pub adapter: Ident,
    pub filename: LitStr,
    pub srcpath: Path,
    pub modpath: Option<Path>,
    pub target: Type,
}

impl Parse for ClientArgs {
    fn parse(input: ParseStream) -> syn::Result<Self> {
        // impl "path"::mod::trait ...
        input.parse::<Token![impl]>()?;
        let filename: LitStr = input.parse()?;
        input.parse::<Token![::]>()?;
        let srcpath: Path = input.parse()?;
        // [as mod::trait] ...
        let lookahead = input.lookahead1();
        let modpath: Option<Path> = if lookahead.peek(Token![as]) {
            input.parse::<Token![as]>()?;
            Some(input.parse()?)
        } else {
            None
        };
        // for Type ...
        input.parse::<Token![for]>()?;
        let target: Type = input.parse()?;
        // in adapter
        input.parse::<Token![in]>()?;
        let adapter: Ident = input.parse()?;
        // )
        if !input.is_empty() {
            return Err(input.error("Wrong syntax for make_client: nothing should be following the struct"));
        }
        Ok(Self {
            adapter,
            filename,
            srcpath,
            modpath,
            target,
        })
    }
}

pub fn make_client_impl(args: TokenStream) -> TokenStream {
    let ClientArgs {
        adapter,
        filename,
        srcpath,
        modpath,
        target,
    } = match syn::parse2(args) {
        Ok(v) => v,
        Err(err) => return err.to_compile_error(),
    };

    // Generate the injecting content
    let content = match read_trait(filename, &srcpath) {
        Some(v) => v,
        None => return TokenStream::new(),
    }
    .iter()
    .map(|item| match item {
        TraitItem::Macro(_) => panic!("macros in trait declaration are not currently supported"),
        value => quote! { #adapter !(#value); },
    })
    .collect::<Vec<_>>();
    let modpath = modpath.unwrap_or(srcpath);

    quote! {
        impl #modpath for #target {
            #(#content)*
        }
    }
}
