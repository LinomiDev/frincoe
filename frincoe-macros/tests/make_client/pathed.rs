use frincoe_macros::make_client;

include!("interfaces.rs");

macro_rules! defaulting {
    ( fn $func:ident $args:tt -> $ret:ty; ) => {
        fn $func $args -> $ret {
            Default::default()
        }
    };
}

struct Target {
}

make_client!(impl "interfaces.rs"::pathing::MoreFns for Target in defaulting);

fn main() {
    use pathing::MoreFns;
    let t = Target {};
    t.f1(1);
    t.f2(123);
    t.f3(1.0, 1.0);
}
