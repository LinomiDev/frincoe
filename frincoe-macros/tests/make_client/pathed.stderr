error[E0382]: borrow of moved value: `t`
  --> tests/make_client/pathed.rs:22:5
   |
20 |     let t = Target {};
   |         - move occurs because `t` has type `Target`, which does not implement the `Copy` trait
21 |     t.f1(1);
   |       ----- `t` moved due to this method call
22 |     t.f2(123);
   |     ^^^^^^^^^ value borrowed here after move
   |
note: this function takes ownership of the receiver `self`, which moves `t`
  --> tests/make_client/interfaces.rs
   |
   |         fn f1(self, x: i32) -> u32;
   |               ^^^^

error[E0596]: cannot borrow `t` as mutable, as it is not declared as mutable
  --> tests/make_client/pathed.rs:23:5
   |
20 |     let t = Target {};
   |         - help: consider changing this to be mutable: `mut t`
...
23 |     t.f3(1.0, 1.0);
   |     ^^^^^^^^^^^^^^ cannot borrow as mutable
